File: contextforge/app/schemas/knowledge.py
"""
Knowledge item request/response schemas.
"""

from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field

from app.models.enums import KnowledgeType, KnowledgeStatus, Visibility


# ==================== List/Filter Parameters ====================

class KnowledgeListParams(BaseModel):
    """Parameters for listing knowledge items."""
    
    knowledge_type: Optional[KnowledgeType] = None
    status: Optional[KnowledgeStatus] = None
    visibility: Optional[Visibility] = None
    tags: Optional[List[str]] = None
    category_id: Optional[int] = None
    search: Optional[str] = None
    page: int = Field(default=1, ge=1)
    limit: int = Field(default=20, ge=1, le=100)


# ==================== Request Schemas ====================

class KnowledgeItemCreate(BaseModel):
    """Request schema for creating a knowledge item."""
    
    knowledge_type: KnowledgeType
    category_id: Optional[int] = None
    title: str = Field(..., min_length=1, max_length=500)
    summary: Optional[str] = None
    content: Dict[str, Any]
    tags: Optional[List[str]] = None
    visibility: Optional[Visibility] = Visibility.INTERNAL
    status: Optional[KnowledgeStatus] = KnowledgeStatus.DRAFT
    metadata_: Optional[Dict[str, Any]] = None
    
    class Config:
        json_schema_extra = {
            "example": {
                "knowledge_type": "faq",
                "title": "How to create a purchase order",
                "content": {
                    "question": "How do I create a new purchase order?",
                    "answer": "Navigate to Purchasing > Create PO..."
                },
                "tags": ["purchasing", "how-to"],
                "visibility": "internal",
                "status": "draft"
            }
        }


class KnowledgeItemUpdate(BaseModel):
    """Request schema for updating a knowledge item."""
    
    category_id: Optional[int] = None
    title: Optional[str] = Field(None, min_length=1, max_length=500)
    summary: Optional[str] = None
    content: Optional[Dict[str, Any]] = None
    tags: Optional[List[str]] = None
    visibility: Optional[Visibility] = None
    status: Optional[KnowledgeStatus] = None
    metadata_: Optional[Dict[str, Any]] = None


# ==================== Response Schemas ====================

class KnowledgeItemResponse(BaseModel):
    """Response schema for a knowledge item."""
    
    id: int
    knowledge_type: KnowledgeType
    category_id: Optional[int] = None
    title: str
    summary: Optional[str] = None
    content: Dict[str, Any]
    tags: List[str]
    visibility: Visibility
    status: KnowledgeStatus
    metadata_: Optional[Dict[str, Any]] = None
    created_by: Optional[str] = None
    created_at: datetime
    updated_by: Optional[str] = None
    updated_at: Optional[datetime] = None
    
    # Extended info (populated on demand)
    variants_count: Optional[int] = None
    relationships_count: Optional[int] = None
    
    class Config:
        from_attributes = True


class KnowledgeItemDetailResponse(KnowledgeItemResponse):
    """Detailed response including variants and relationships."""
    
    variants: Optional[List["VariantResponse"]] = None
    related_items: Optional[List["RelatedItemResponse"]] = None


# ==================== Variant Schemas ====================

class VariantCreate(BaseModel):
    """Request schema for creating a variant."""
    
    variant_text: str = Field(..., min_length=1)
    source: Optional[str] = "manual"
    source_reference: Optional[str] = None


class VariantResponse(BaseModel):
    """Response schema for a variant."""
    
    id: int
    knowledge_item_id: int
    variant_text: str
    source: str
    source_reference: Optional[str] = None
    created_by: Optional[str] = None
    created_at: datetime
    
    class Config:
        from_attributes = True


# ==================== Relationship Schemas ====================

class RelationshipCreate(BaseModel):
    """Request schema for creating a relationship."""
    
    target_id: int
    relationship_type: str
    weight: Optional[float] = 1.0
    is_bidirectional: Optional[bool] = False


class RelationshipResponse(BaseModel):
    """Response schema for a relationship."""
    
    id: int
    source_id: int
    target_id: int
    relationship_type: str
    weight: float
    is_bidirectional: bool
    is_auto_generated: bool
    created_by: Optional[str] = None
    created_at: datetime
    
    class Config:
        from_attributes = True


class RelatedItemResponse(BaseModel):
    """Response schema for related items (includes relationship info)."""
    
    id: int
    knowledge_type: KnowledgeType
    title: str
    relationship_type: str
    is_auto_generated: bool


# ==================== Category Schemas ====================

class CategoryCreate(BaseModel):
    """Request schema for creating a category."""
    
    name: str = Field(..., min_length=1, max_length=100)
    slug: str = Field(..., min_length=1, max_length=100, pattern=r"^[a-z0-9-]+$")
    description: Optional[str] = None
    parent_id: Optional[int] = None
    default_visibility: Optional[str] = "internal"
    sort_order: Optional[int] = 0


class CategoryUpdate(BaseModel):
    """Request schema for updating a category."""
    
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = None
    parent_id: Optional[int] = None
    default_visibility: Optional[str] = None
    sort_order: Optional[int] = None


class CategoryResponse(BaseModel):
    """Response schema for a category."""
    
    id: int
    name: str
    slug: str
    description: Optional[str] = None
    parent_id: Optional[int] = None
    default_visibility: str
    sort_order: int
    created_at: datetime
    
    class Config:
        from_attributes = True


class CategoryTreeResponse(CategoryResponse):
    """Response schema for category with children."""
    
    children: Optional[List["CategoryTreeResponse"]] = None


# ==================== Version Schemas ====================

class VersionResponse(BaseModel):
    """Response schema for a version."""
    
    id: int
    knowledge_item_id: int
    version_number: int
    title: str
    summary: Optional[str] = None
    content: Dict[str, Any]
    tags: Optional[List[str]] = None
    change_type: Optional[str] = None
    change_reason: Optional[str] = None
    changed_by: Optional[str] = None
    changed_at: datetime
    
    class Config:
        from_attributes = True


# Forward references
KnowledgeItemDetailResponse.model_rebuild()
CategoryTreeResponse.model_rebuild()

File: admin-ui/src/types/knowledge.ts
import { z } from 'zod'

// ==================== Knowledge Types ====================

export const KnowledgeType = {
  FAQ: 'faq',
  PERMISSION: 'permission',
  SCHEMA: 'schema',
  EXAMPLE: 'example',
  PLAYBOOK: 'playbook',
} as const

export type KnowledgeType = (typeof KnowledgeType)[keyof typeof KnowledgeType]

export const KnowledgeStatus = {
  DRAFT: 'draft',
  PUBLISHED: 'published',
  ARCHIVED: 'archived',
} as const

export type KnowledgeStatus = (typeof KnowledgeStatus)[keyof typeof KnowledgeStatus]

export const Visibility = {
  PUBLIC: 'public',
  INTERNAL: 'internal',
  RESTRICTED: 'restricted',
} as const

export type Visibility = (typeof Visibility)[keyof typeof Visibility]

export const ExampleFormat = {
  TEXT: 'text',
  JSON: 'json',
} as const

export type ExampleFormat = (typeof ExampleFormat)[keyof typeof ExampleFormat]

// ==================== Content Types ====================

export interface FAQContent {
  answer: string // Markdown supported
}

export interface PermissionContent {
  description: string
  permissions: string[]
  roles: string[]
  context?: string
}

export interface SchemaContent {
  name: string
  definition: string // YAML string (YAMLSchemaV1 format)
}

export interface ExampleContent {
  schema_id: number
  description: string
  content: string // JSON or text based on format
  format: ExampleFormat
}

export interface PlaybookContent {
  domain: string
  content: string // Markdown supported
}

export type KnowledgeContent =
  | FAQContent
  | PermissionContent
  | SchemaContent
  | ExampleContent
  | PlaybookContent

// ==================== Base Knowledge Item ====================

export interface BaseKnowledgeItem {
  id: number
  knowledge_type: KnowledgeType
  title: string
  tags: string[]
  status: KnowledgeStatus
  visibility: Visibility
  created_by?: string
  created_at: string
  updated_by?: string
  updated_at?: string
}

// ==================== Type-Specific Items ====================

export interface FAQItem extends BaseKnowledgeItem {
  knowledge_type: typeof KnowledgeType.FAQ
  content: FAQContent
}

export interface PermissionItem extends BaseKnowledgeItem {
  knowledge_type: typeof KnowledgeType.PERMISSION
  content: PermissionContent
}

export interface SchemaItem extends BaseKnowledgeItem {
  knowledge_type: typeof KnowledgeType.SCHEMA
  content: SchemaContent
}

export interface ExampleItem extends BaseKnowledgeItem {
  knowledge_type: typeof KnowledgeType.EXAMPLE
  content: ExampleContent
  // Populated field
  schema?: SchemaItem
}

export interface PlaybookItem extends BaseKnowledgeItem {
  knowledge_type: typeof KnowledgeType.PLAYBOOK
  content: PlaybookContent
}

export type KnowledgeItem =
  | FAQItem
  | PermissionItem
  | SchemaItem
  | ExampleItem
  | PlaybookItem

// ==================== Domain Management ====================

export interface Domain {
  id: string
  name: string
  description?: string
  isCustom: boolean // User-added vs predefined
}

export const DEFAULT_DOMAINS: Domain[] = [
  { id: 'purchasing', name: 'Purchasing', description: 'Purchase orders and procurement', isCustom: false },
  { id: 'sales', name: 'Sales', description: 'Sales orders and customer management', isCustom: false },
  { id: 'inventory', name: 'Inventory', description: 'Stock and warehouse management', isCustom: false },
  { id: 'finance', name: 'Finance', description: 'Financial operations and reporting', isCustom: false },
  { id: 'hr', name: 'Human Resources', description: 'Employee and HR management', isCustom: false },
  { id: 'analytics', name: 'Analytics', description: 'Data analytics and reporting', isCustom: false },
]

// ==================== Zod Schemas for Forms ====================

export const faqFormSchema = z.object({
  title: z.string().min(10, 'Question must be at least 10 characters'),
  answer: z.string().min(20, 'Answer must be at least 20 characters'),
  tags: z.array(z.string()),
  status: z.enum(['draft', 'published', 'archived']),
  visibility: z.enum(['public', 'internal', 'restricted']),
})

export type FAQFormData = z.infer<typeof faqFormSchema>

export const permissionFormSchema = z.object({
  title: z.string().min(3, 'Feature name must be at least 3 characters'),
  description: z.string().min(10, 'Description must be at least 10 characters'),
  permissions: z.array(z.string()).min(1, 'At least one permission is required'),
  roles: z.array(z.string()).min(1, 'At least one role is required'),
  context: z.string().optional(),
  tags: z.array(z.string()),
  status: z.enum(['draft', 'published', 'archived']),
  visibility: z.enum(['public', 'internal', 'restricted']),
})

export type PermissionFormData = z.infer<typeof permissionFormSchema>

export const schemaFormSchema = z.object({
  title: z.string().min(3, 'Title must be at least 3 characters'),
  name: z.string().min(1, 'Schema name is required').regex(/^[a-z0-9_-]+$/, 'Name must be lowercase alphanumeric with dashes/underscores'),
  definition: z.string().min(10, 'Schema definition is required'),
  tags: z.array(z.string()),
  status: z.enum(['draft', 'published', 'archived']),
  visibility: z.enum(['public', 'internal', 'restricted']),
})

export type SchemaFormData = z.infer<typeof schemaFormSchema>

export const exampleFormSchema = z.object({
  title: z.string().min(3, 'Title must be at least 3 characters'),
  schema_id: z.number().min(1, 'Schema is required'),
  description: z.string().min(10, 'Description must be at least 10 characters'),
  content: z.string().min(1, 'Content is required'),
  format: z.enum(['text', 'json']),
  tags: z.array(z.string()),
  status: z.enum(['draft', 'published', 'archived']),
  visibility: z.enum(['public', 'internal', 'restricted']),
})

export type ExampleFormData = z.infer<typeof exampleFormSchema>

export const playbookFormSchema = z.object({
  title: z.string().min(3, 'Title must be at least 3 characters'),
  domain: z.string().min(1, 'Domain is required'),
  content: z.string().min(20, 'Content must be at least 20 characters'),
  tags: z.array(z.string()),
  status: z.enum(['draft', 'published', 'archived']),
  visibility: z.enum(['public', 'internal', 'restricted']),
})

export type PlaybookFormData = z.infer<typeof playbookFormSchema>

// ==================== API Types ====================

export interface PaginatedResponse<T> {
  data: T[]
  total: number
  page: number
  limit: number
  total_pages: number
}

export interface KnowledgeFilters {
  search?: string
  knowledge_type?: KnowledgeType[]
  tags?: string[]
  status?: KnowledgeStatus[]
  visibility?: Visibility
  domain?: string // For playbooks
  schema_id?: number // For examples
}

// ==================== Metrics Types ====================

export interface TypeStats {
  type: KnowledgeType
  total: number
  published: number
  draft: number
  archived: number
}

export interface KnowledgeMetrics {
  total_items: number
  by_type: TypeStats[]
  by_status: Record<KnowledgeStatus, number>
  recent_activity: {
    date: string
    created: number
    updated: number
  }[]
}

export interface KnowledgeHitStats {
  id: number
  knowledge_type: KnowledgeType
  title: string
  tags: string[]
  total_hits: number
  unique_sessions: number
  days_with_hits: number
  last_hit_at: string
  avg_similarity: number
  primary_retrieval_method: string
}

export interface DailyHitStats {
  date: string
  total_hits: number
  unique_sessions: number
  by_type: Record<KnowledgeType, number>
}

// ==================== Legacy Types (for backward compatibility) ====================

export interface LegacyKnowledgeItem {
  id: number
  knowledge_type: string
  category_id?: number
  category?: KnowledgeCategory
  title: string
  summary?: string
  content: Record<string, unknown>
  tags: string[]
  visibility: string
  status: string
  owner_id?: string
  team_id?: string
  created_by?: string
  created_at: string
  updated_at?: string
  variants_count?: number
  relationships_count?: number
  hits_count?: number
}

export interface KnowledgeCategory {
  id: number
  name: string
  slug: string
  description?: string
  parent_id?: number
  default_visibility: string
  sort_order: number
}

export interface KnowledgeVariant {
  id: number
  knowledge_item_id: number
  variant_text: string
  source: string
  source_reference?: string
  created_by?: string
  created_at: string
}

export interface KnowledgeRelationship {
  id: number
  source_id: number
  target_id: number
  relationship_type: string
  weight: number
  is_bidirectional: boolean
  is_auto_generated: boolean
  created_by?: string
  created_at: string
}

export interface StagingKnowledgeItem {
  id: number
  node_type: string
  tenant_id: string
  title: string
  content: Record<string, unknown>
  tags: string[]
  status: 'pending' | 'approved' | 'rejected'
  action: 'new' | 'merge' | 'add_variant'
  merge_with_id?: number
  similarity?: number
  source?: string
  source_reference?: string
  confidence?: number
  created_at: string
  created_by?: string
  reviewed_by?: string
  reviewed_at?: string
  review_notes?: string
}
